.pre-commit-config.yaml

```diff
@@ -47,46 +19,32 @@ repos:
       - id: trailing-whitespace
 
   - repo: https://github.com/pre-commit/mirrors-mypy
-    rev: "v0.991"
+    rev: "v1.1.1"
     hooks:
       - id: mypy
         # Do not install *types-click* - it's not recommended with Click 8 & newer,
         # which is the version a developer encounters given the requirements are not
         # frozen.
-        additional_dependencies: [click]
+        #additional_dependencies: [click]
         pass_filenames: false
-        args:
-          - --config-file
-          - mypy.ini
+        language: system
```

Comment:
Spent 2 hours trying to make pre-commit mypy pass with --all-files. This is the least insane way I found that should work consistently with devs running `mypy` command, pre-commit hooks when doing git commit and CI running --all-files.
Note I've added python_version=3.7 to mypy config.

Reply:
I believe Python 3.6 is still required, because of RHEL8. Not sure what effect it has on mypy, though.

But yes, making all use cases happy is tricky, I discovered a different set of issues reported when running mypy via `pre-commit` vs. running it directly. So it is tricky.

Reply:
> I discovered a different set of issues reported when running mypy via pre-commit vs. running it directly. So it is tricky.

Indeed, that's why locking the version + language:system was needed, as pre-commit creates virtualenv with (most likely) different python version.

> I believe Python 3.6 is still required

A lot of the tools do not work with 3.6 as it's EOL. I'd say it's best to lock everything to py37 instead of having some on py36 and some on py37? (el8 users can install python39 btw)

Reply:
I know it's EOL :) Yet, it's still with us... Anyway, what exactly is `python_version` for? Is it describing the Python powering mypy, or the desired Python version against which we want the project checked? limit it in this way? I've read the documentation, it sounds like the former, do we have to set it/enforce it? Github action seems to use Python 3.11, would this lead to a mismatch?

Reply:
Yes I believe it's the former. 
There is a mismatch with default pre-commit virtual environment which pre-commit is creating for the mypy hook. It produces results like if it's either running <=3.6 or has different target? :shrug: 
I solved it by using `language: system`. This tells it to use system's interpreter, so for example in CI, it requires `actions/setup-python` which is already being run anyway.

Alternatively, we could try `language_version: python3.X` instead of `language: system`, however this does not work well with `install_types = true` as far as understand.

Reply:
```
 mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

Executable `mypy` not found
```

Oh, should have seen that coming.
I was thinking - maybe the `language: system` isn't that great idea, if only for the potential issues with different pythons in CI and individual dev environments. Now I'm trying to make it work without `--instal-types` and have this:
```
  - repo: https://github.com/pre-commit/mirrors-mypy
    rev: "v1.2.0"
    hooks:
      - id: mypy
        additional_dependencies:
          - click
          - types-Markdown
          - types-requests
          - types-setuptools
          - types-urllib3
          - ruamel.yaml

        pass_filenames: false
        args: [--config-file=pyproject.toml]
```
But there are two things that I don't understand:

1) Without ruamel as dependency(and in 'ignore-imports' list), it does not like these two ignores:
```
tmt/utils.py:1594: error: Unused "type: ignore" comment
tmt/utils.py:1595: error: Unused "type: ignore" comment
```
while with it installed they are needed. But fair enough, we can do either/or.  


2) This error is only caught with pre-commit, not when I run mypy manually:  
`tmt/utils.py:4377: error: Returning Any from function declared to return "str"  [no-any-return]`

I assume the error is about utils.render_template_file() not return (empty) `str` if `environment.from_string(template_filepath.read_text())` ?
Why is it different from when I run `mypy` from command line, even when it's also python3.11 and uses same pyproject.toml config?